### ISSUE TITLE
validation of response schema not working for recursive schemas

### ISSUE THREAD

**Issue by jeremyschiemann** [kiwicom/schemathesis#468]
In `response_schema_conformance` (in checks.py) schemathesis is validating the response schema after it has tried to resolve it. 
If the response schema is recursive this doesnt work. It would be better to directly pass the response schema to `jsonschema.validate()` instead of the resolved one.

if i replace
`schema = case.endpoint.schema._get_response_schema(definition)`
with
`schema = responses[str(response.status_code)]['content'][content_type]` (will fail if response has no content type)
the validation works fine. 

since `jsonschema.validate()` can handle recursive schemas, this might also be a better fix for https://github.com/kiwicom/schemathesis/issues/435, validate the raw schema and then resolve it as much as needed to safe time wasted in unnecessary recursions.

PS: just wanted to say thank you for schemathesis and not just keep opening new issues. I t has really helped us ;)

**Comment by Stranger6667** [comment]
Interesting case, could you please share some small schema that reproduces the issue?

> since jsonschema.validate() can handle recursive schemas, this might also be a better fix for #435, validate the raw schema and then resolve it as much as needed to safe time wasted in unnecessary recursions.

Unfortunately, it won't work on `hypothesis-jsonschema` level, that is explicitly doesn't support recursive schemas :(

> PS: just wanted to say thank you for schemathesis and not just keep opening new issues. I t has really helped us ;)

I am glad that Schemathesis helped you :) And I appreciate all reported issues, because they help to improve the project :)

**Comment by jeremyschiemann** [comment]
I extracted [these 2 components](https://github.com/kiwicom/schemathesis/files/4427529/extracted.txt) out of our components/schemas and reduced them to only the necessary bits.

in our endpoint we referenced this for a 200 response like so
```
"content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/eventItemResponse"
                }
              }
            }
```
so eventTree is referencing itself. Does this help you?

**Comment by Stranger6667** [comment]
Yep, it definitely helps, thank you :) I'll take a deeper look a bit later

**Comment by jeremyschiemann** [comment]
thank you very much!

**Comment by Stranger6667** [comment]
I did some investigation on the issue and here are my observations.

1. With recursive schemas `case.endpoint.definition` contains **partially** resolved schemas until some limit.
2. And this reference in the final node has no way to be resolved since other parts of the document are not in this variable (e.g. there is no path `"#/components/schemas/Node"` in this document - it exists only in the full schema)
3. If the response is nested enough to get to this reference it will fail during resolution

I assume that it is the error you encountered. Please, share your thoughts if I correctly identified this problem.

Regarding solution. I think that you proposed the right thing - using the raw schema definitely should solve it.

I think that it could look like this:

```
jsonschema.validate(response, response_schema, resolver=raw_schema_resolver)
```

So, actually we can either put not resolved schemas to `definition` and add the resolver from the base schema or we can only add resolver (it will resolve the ref when it will be encountered)

I'll try to play with these ideas and will share the progress

**Comment by Stranger6667** [comment]
The fix is released in `1.1.0` - https://pypi.org/project/schemathesis/1.1.0/

Please, let me know if it fixes your use-case

**Comment by jeremyschiemann** [comment]
> Please, share your thoughts if I correctly identified this problem.

thats exactly what i was thinking about this problem

> Please, let me know if it fixes your use-case

 sadly it doesnt, 

the `resolver=case.endpoint.schema.resolver` you added in https://github.com/kiwicom/schemathesis/pull/484/commits/756f3ddc4b4c08ca90ab0daafcdf1347c8d88dd0 is firstly used after my error occures.

```python
def validate(instance, schema, cls=None, *args, **kwargs):
    if cls is None:
        cls = validator_for(schema)

    cls.check_schema(schema)
    validator = cls(schema, *args, **kwargs)
    error = exceptions.best_match(validator.iter_errors(instance))
    if error is not None:
        raise error
```

this is the jsonschema.validate() function, my error occurs in `cls.check_schema(schema)` not in the row after where the kwargs are used.

> So, actually we can either put not resolved schemas to definition and add the resolver from the base schema

maybe this would be the better approach?

**Comment by Stranger6667** [comment]
Oh, I see. Could you please, share the error traceback here?

**Comment by Stranger6667** [comment]
I have a guess - the error might happen if the schema doesn't conform to Draft7 (which is the default in `jsonschema`), e.g. if there is something specific to Open API or from Draft4, for example, `exclusiveMinimum`. I'll try to reproduce this idea tomorrow

**Comment by Stranger6667** [comment]
> maybe this would be the better approach?

Yep, I will work on this - it requires some extra copying & handing [here](https://github.com/kiwicom/schemathesis/blob/master/src/schemathesis/schemas.py#L212)

**Comment by Stranger6667** [comment]
@jeremyschiemann 

My assumption from the previous message was correct, and the fix is released in `1.1.1` (https://pypi.org/project/schemathesis/1.1.1/). It might be the cause of the error you mentioned above (the one in `cls.check_schema(schema)`) - I reproduced a schema error on this line in the test modified in #490 

After this fix, it becomes harder to pass not resolved schema to the `endpoint`, but I will work on it anyway.

Let me know if `1.1.1` fixes your case. Sorry for inconvenience

**Comment by jeremyschiemann** [comment]
hi @Stranger6667  im on on schemathesis 1.2.0 now and i still need to apply the fix from the very first message.

I will try to find something that will help you fix it or see if I can create a schema which reproduces my error when I find the time

**Comment by Stranger6667** [comment]
Hi @jeremyschiemann !
I am very sorry to hear that :( Thank you for your patience

**Comment by Stranger6667** [comment]
The fix is released in 1.3.0. Now those definitions are not resolved which fixes the recursion error (I was able to reproduce it)
https://pypi.org/project/schemathesis/1.3.0/

Let me know if it fixes it in your case as well

**Comment by jeremyschiemann** [comment]
Hi again,

thanks for fixing it, it works now :)
